### PR TITLE
Test operator

### PR DESCRIPTION
**Pull request by peter-tackage**
All done from my side!

@tomaszpolanski Should the version be incremented in this PR too?

### CODE REVIEW

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    public OptionAssertion<T> assertIsSome() {
+        if (!actual.isSome()) {
+            fail("Option was not Some");
+        }
+        return this;
+    }
+
+    public OptionAssertion<T> assertValue(Func1<T, Boolean> predicate) {
```
Comment (peter-tackage) [inline options/core/src/main/java/polanski/option/OptionAssertion.java]: I should add null checks for the predicate.

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            fail("Option was not Some");
+        }
+
+        if (!actual.filter(predicate).isSome()) {
+            fail("Option value did not match predicate");
+        }
+        return this;
+    }
+
+    public OptionAssertion<T> assertValue(T expected) {
```
Comment (peter-tackage) [inline options/core/src/main/java/polanski/option/OptionAssertion.java]: I should add null checks for the value - I guess? I mean the value could be null, but the test should use `assertNone` instead.

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import android.support.annotation.NonNull;
+
+import polanski.option.function.Func1;
+
+public final class OptionAssertion<T> {
+
+    @NonNull
+    private final Option<T> actual;
+
+    OptionAssertion(@NonNull final Option<T> actual) {
```
Comment (peter-tackage) [inline options/core/src/main/java/polanski/option/OptionAssertion.java:18]: So long as we can only create this via `Option`, then we can be sure that this is non-null.

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        this.actual = actual;
+    }
+
+    public void assertIsNone() {
+        if (!actual.isNone()) {
+            fail("Option was not None");
+        }
+    }
```
Comment (peter-tackage) [inline options/core/src/main/java/polanski/option/OptionAssertion.java:31]: I don't chain this one, because I didn't think that anything else could be useful chained afterwards, but happy to make it return `OptionAssertion<T>` if you want.

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+        if (!actual.filter(v -> v.equals(expected)).isSome()) {
+            fail(String.format("Option value: <%s> did not equal expected value: <%s>",
+                               OptionUnsafe.getUnsafe(actual),
+                               expected));
+        }
+        return this;
+    }
+
+    private void fail(String message) {
```
Comment (peter-tackage) [inline options/core/src/main/java/polanski/option/OptionAssertion.java]: Can you think of any other useful "debug" info to add to the failure message?

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
+package polanski.option;
+
+import android.support.annotation.NonNull;
+
+import polanski.option.function.Func1;
+
+public final class OptionAssertion<T> {
```
Comment (peter-tackage) [inline options/core/src/main/java/polanski/option/OptionAssertion.java:13]: Perhaps we could also expose some predicates `isNone` etc to make testing with Rx more streamlined. Although, I guess that wouldn't really work because it would require this library to use Rx predicate function classes.

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            fail(String.format("Option value: <%s> did not equal expected value: <%s>",
+                               OptionUnsafe.getUnsafe(actual),
+                               expected));
+        }
+        return this;
+    }
+
+    @NonNull
+    private static <T> Func1<T, Boolean> equalsPredicate(@NonNull final T expected) {
+        return v -> v.equals(expected);
```
Comment (peter-tackage) [inline options/core/src/main/java/polanski/option/OptionAssertion.java]: Should this be a lambda or just use a `Func1` explicitly?

File: options/core/src/test/java/polanski/option/OptionAssertionTest.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Test
+    public void assertValue_throwsAssertionError_whenOptionNone_notEqualTo() {
+        thrown.expect(AssertionError.class);
+        thrown.expectMessage(startsWith("Option was not Some"));
+
+        new OptionAssertion<>(Option.none())
+                .assertValue("expected");
+    }
+
```
Comment (peter-tackage) [inline options/core/src/test/java/polanski/option/OptionAssertionTest.java:119]: I should add tests for the null precondition checks too.

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        return this;
+    }
+
+    @NonNull
+    public OptionAssertion<T> assertValue(@NonNull final Func1<T, Boolean> predicate) {
+        checkNotNull(predicate, "Predicate function cannot be null");
+
+        if (!actual.isSome()) {
+            fail("Option was not Some");
```
Comment (tomaszpolanski) [inline options/core/src/main/java/polanski/option/OptionAssertion.java]: How about "Option was None"?

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        }
+
+        if (!matches(actual, predicate)) {
+            fail("Option value did not match predicate");
```
Comment (tomaszpolanski) [inline options/core/src/main/java/polanski/option/OptionAssertion.java]: How about printing the value?

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        return this;
+    }
+
+    @NonNull
+    public OptionAssertion<T> assertValue(@NonNull final T expected) {
+        checkNotNull(expected, "Expected value cannot be null: use assertNone instead");
+
+        if (!actual.isSome()) {
+            fail("Option was not Some");
```
Comment (tomaszpolanski) [inline options/core/src/main/java/polanski/option/OptionAssertion.java]: Same as above

File: options/core/src/main/java/polanski/option/OptionAssertion.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    private static <T> Func1<T, Boolean> equalsPredicate(@NonNull final T expected) {
+        return v -> v.equals(expected);
+    }
+
+    private static <T> boolean matches(@NonNull final Option<T> actual,
+                                       @NonNull final Func1<T, Boolean> predicate) {
+        return actual.filter(predicate).isSome();
+    }
+
+    private void fail(String message) {
```
Comment (tomaszpolanski) [inline options/core/src/main/java/polanski/option/OptionAssertion.java]: NonNull?

File: options/core/src/test/java/polanski/option/OptionAssertionTest.java (hunk 8 above)
Comment (tomaszpolanski) [inline options/core/src/test/java/polanski/option/OptionAssertionTest.java]: There is not need

**Review by tomaszpolanski — CHANGES_REQUESTED** [review]

File: options/core/src/main/java/polanski/option/OptionAssertion.java (hunk 9 above)
Comment (peter-tackage) [reply options/core/src/main/java/polanski/option/OptionAssertion.java]: I was thinking that it's better to say that the condition that they are testing for *is not* true and then assert, rather saying that some other condition *is* true and then assert.

File: options/core/src/main/java/polanski/option/OptionAssertion.java (hunk 12 above)
Comment (peter-tackage) [reply options/core/src/main/java/polanski/option/OptionAssertion.java]: I didn't annotate it because it is an internal private method and not subject to external abuse . It also supports nullability, so if anything it should `@Nullable`. Thoughts?

File: options/core/src/main/java/polanski/option/OptionAssertion.java (hunk 10 above)
Comment (peter-tackage) [reply options/core/src/main/java/polanski/option/OptionAssertion.java]: The `fail` method does this by default. I could make it a little more readable though.

File: options/core/src/main/java/polanski/option/OptionAssertion.java (hunk 7 above)
Comment (tomaszpolanski) [reply options/core/src/main/java/polanski/option/OptionAssertion.java]: I do not see a reason why it should be a Func1. Static method should be enough

File: options/core/src/main/java/polanski/option/OptionAssertion.java (hunk 12 above)
Comment (tomaszpolanski) [reply options/core/src/main/java/polanski/option/OptionAssertion.java]: Some private methods could also accept nullable arguments so I would prefer to have annotation as well here

**Comment by tomaszpolanski** [comment]
@peter-tackage I've added some final comments

File: options/core/src/main/java/polanski/option/OptionAssertion.java (hunk 7 above)
Comment (peter-tackage) [reply options/core/src/main/java/polanski/option/OptionAssertion.java]: Java 6/7 compatibility? To be honest, I'm not sure why it is allowing the lambda at the moment; shouldn't it fail? There's no retrolambda. Are we building with the correct `sourceCompatibility`?

Comment (peter-tackage) [reply options/core/src/main/java/polanski/option/OptionAssertion.java]: A statically defined `Func1` "equals" instance would be very slightly faster than creating a new instance each time that the predicate is run.

File: options/core/src/test/java/polanski/option/OptionAssertionTest.java (hunk 8 above)
Comment (peter-tackage) [reply options/core/src/test/java/polanski/option/OptionAssertionTest.java:119]: I've done this, just for the sake of documenting it.

**Review by tomaszpolanski — APPROVED** [review]
